File: webapp/src/app/representation/annotation.ts
import {AtomicContainer, Sequence} from "./container";
import {Annotation} from "./scenario";

enum ImageLabel {
  FACE
}

export enum EntityType {
  PERSON, FRIEND, OBJECT = 2
}

export interface Entity {
  id: string;
  type: EntityType;
}

export interface Triple {
  subject: Entity
  predicate: string;
  object: Entity
}

export interface Token extends AtomicContainer<string> {
  // no additional fields
}

export interface Utterance extends Sequence<Token> {
  chat_id: string;
  utterance: string;
  tokens: Token[]
}

export interface Display {
  display: string;
}

export function annotationDisplayValue(annotation: Annotation<any>): string {
  switch (annotation.type.toLowerCase()) {
    case "display":
    case "conversationalagent":
    case "label":
    case "pos":
    case "emotion":
      return annotation.value;
    case "triple":
      return annotation.value.subject.id + annotation.value.predicate + annotation.value.object.id;
    case "utterance":
      return annotation.value.utterance;
    case "entity":
      return annotation.value.id + " - " + annotation.value.type;
    case "person":
      return annotation.value.name;
    case "object":
      return annotation.value.label;
    case "token":
      return annotation.value.value || annotation.value;
    default:
      if (annotation.value !== Object(annotation.value)) {
        // pass
      } else if("display" in annotation.value && annotation.value.display) {
        return annotation.value.display;
      } else if ("name" in annotation.value && annotation.value.name) {
        return annotation.value.name;
      } else if ("label" in annotation.value && annotation.value.label) {
        return annotation.value.label;
      }

      return annotation.type.startsWith("python-type:") ?
          annotation.type.split('.').slice(-1)[0] :
          annotation.type;
  }
}

File: emissor/representation/annotation.py
# TODO This conflicts with marshmallow dataclass serialization
# Enable using the current class in type annotations
# from __future__ import annotations

"""
Basic classes for values used in :class:`emissor.representation.Annotation`.
"""
from dataclasses import dataclass
from enum import Enum, auto

import uuid
from rdflib import URIRef, Namespace
from typing import Any, List

from emissor.representation.container import Sequence, AtomicContainer, AtomicRuler
from emissor.representation.ldschema import LdId, emissor_dataclass
from emissor.representation.util import Identifier

friends_namespace = Namespace("http://cltl.nl/leolani/friends/")
data_namespace = Namespace("http://cltl.nl/combot/signal/")
predicate_namespace = Namespace("http://cltl.nl/combot/predicate/")


class AnnotationType(Enum):
    DISPLAY = auto()
    PERSON = auto()
    EMOTION = auto()
    FRIEND = auto()
    OBJECT = auto()
    TOKEN = auto()
    POS = auto()
    NER = auto()
    LINK = auto()
    REPRESENTATION = auto()
    UTTERANCE = auto()


class ImageLabel(Enum):
    FACE = auto()


class EntityType(Enum):
    PERSON = auto()
    FRIEND = auto()
    OBJECT = auto()


@emissor_dataclass
class Entity:
    id: URIRef = LdId()


@dataclass
class Token(AtomicContainer[str]):
    @classmethod
    def for_string(cls, value: str):
        token_id = str(uuid.uuid4())
        return cls(token_id, AtomicRuler(token_id), value)


@dataclass
class NER(AtomicContainer[str]):
    @classmethod
    def for_string(cls, value: str):
        token_id = str(uuid.uuid4())
        return cls(token_id, AtomicRuler(token_id), value)


@dataclass
class Triple:
    subject: Entity
    predicate: URIRef
    object: Entity

    # TODO make this more generic, return Triple (see import of annotations)
    @classmethod
    def from_friends(cls, subject_id, predicate_id, object_id) -> Any:
        return cls(Entity(friends_namespace.term(subject_id), EntityType.FRIEND),
                   predicate_namespace.term(predicate_id),
                   Entity(friends_namespace.term(object_id), EntityType.FRIEND))


@dataclass
class Utterance(Sequence):
    chat_id: Identifier
    utterance: str
    tokens: List[Token]


@dataclass
class Display:
    display: str
